docker-storage-setup.sh

```diff
@@ -111,12 +111,48 @@ get_deferred_deletion_string() {
   [ -z "$major" ] && return 0
   [ -z "$minor" ] && return 0
 
-  # docker 1.9 onwards supports deferred device deletion. Enable it.
-  if [ $major -gt 1 ] || ([ $major -eq 1 ] && [ $minor -ge 9 ]);then
-    echo "--storage-opt dm.use_deferred_deletion=true"
+  if should_enable_deferred_deletion $major $minor; then
+     echo "--storage-opt dm.use_deferred_deletion=true"
   fi
 }
 
+should_enable_deferred_deletion() {
+   # docker 1.9 onwards supports deferred device deletion. Enable it.
+   local major=$1
+   local minor=$2
+   if [ $major -lt 1 ] || ([ $major -eq 1 ] && [ $minor -lt 9 ]);then
+      return 1
+   fi
+   if platform_supports_deferred_deletion; then
+      return 0
+   fi
+   return 1
+}
+
+platform_supports_deferred_deletion() {
+        local deferred_deletion_supported=1
+        local pipe1=/tmp/dss-fifo1
+        local pipe2=/tmp/dss-fifo2
+        mkfifo $pipe1
+        mkfifo $pipe2
+        tmpDir=$(mktemp --tmpdir -d)
+        mount -o bind $tmpDir $tmpDir
+        unshare -m ./dss-child-read-write.sh $pipe1 $pipe2 &
+        read n < $pipe1
+        umount $tmpDir
+        rmdir $tmpDir > /dev/null 2>&1
+        deferred_deletion_supported=$?
+        echo "finish" > $pipe2
+        trap cleanup_pipes EXIT
+        return $deferred_deletion_supported
+}
+
+cleanup_pipes(){
+    rm -f /tmp/dss-fifo1
```

Comment:
Let us make /tmp/dss-fifo1 and /tmp/dss-fifo2 global variables and use both in cleanup_pipes and platform_supports_deferred_deletion(). This will avoid hardcoding file name in two functions.

Also let us make tmpDIR global "TEMPDIR".